hoomd/md/external/wall.py

```diff
@@ -0,0 +1,653 @@
+# Copyright (c) 2009-2021 The Regents of the University of Michigan
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
+# License.
+
+r"""Wall potentials.
+
+Wall potentials add forces to any particles within a certain distance,
+:math:`r_{\mathrm{cut}}`, of each wall. In the extrapolated mode, all particles
+deemed outside of the wall boundary are included as well.
+
+Wall geometries (`hoomd.wall`) are used to specify half-spaces. There are two
+half spaces for each of the possible geometries included and each can be
+selected using the inside parameter. In order to fully specify space, it is
+necessary that one half space be closed and one open. Setting *inside=True* for
+closed half-spaces and *inside=False* for open ones. See `hoomd.wall` for more
+information on wall geometries and `WallPotential` for information about forces
+and half-spaces.
+
+.. attention::
+    The current wall force implementation does not support NPT integrators.
+"""
+
+from hoomd.md import force
+from hoomd.data.array_view import _ArrayViewWrapper
+from hoomd.md import _md
+import hoomd
+
+
+def _to_md_cpp_wall(wall):
+    if isinstance(wall, hoomd.wall.Sphere):
+        return _md.SphereWall(wall.radius, wall.origin, wall.inside)
+    if isinstance(wall, hoomd.wall.Cylinder):
+        return _md.CylinderWall(wall.radius, wall.origin, wall.axis,
+                                wall.inside)
+    if isinstance(wall, hoomd.wall.Plane):
+        return _md.PlaneWall(wall.origin, wall.normal, wall.inside)
+    raise TypeError(f"Unknown wall type encountered {type(wall)}.")
+
+
+class _WallArrayViewFactory:
+
+    def __init__(self, cpp_wall_potential, wall_type):
+        self.cpp_obj = cpp_wall_potential
+        self.func_name = {
+            hoomd.wall.Sphere: "get_sphere_list",
+            hoomd.wall.Cylinder: "get_cylinder_list",
+            hoomd.wall.Plane: "get_plane_list"
+        }[wall_type]
+
+    def __call__(self):
+        return getattr(self.cpp_obj.field, self.func_name)()
+
+
+class WallPotential(force.Force):
+    r"""Generic wall potential.
+
+    Warning:
+        `WallPotential` should not be used directly.  It is a base class that
+        provides features and documentation common to all standard wall
+        potentials.
+
+    All wall potential commands specify that a given potential energy and
+    potential be computed on all particles in the system within a cutoff
+    distance, :math:`r_{\mathrm{cut}}`, from each wall in the given wall group.
```

Comment:
```suggestion
    distance, :math:`r_{\mathrm{cut}}`, from each wall.
```

Reply:
Also, reword to account for the extrapolated mode which you mention correctly in the module level docs.